pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,303 @@
+
+#### Decorators
+
+The recommended usage of a decorator is to store metadata about a class or a
+class method. The decorator implementation will usually also provide a function
+to retrieve the metadata stored by it based on the className and method name.
+For a unit test, it is important to test that we can store metadata via a
+decorator and retrieve the correct metadata. *The retrieval gets tested as a
+consequence of validating the metadata was stored.*
+
+**`src/decorators/test.decorator.ts`**
+```ts
+export function test(file: string) {
+  return function(target: Object, methodName: string): void {
+    Reflector.defineMetadata(
+      'example.msg.decorator.metadata.key',
+      {file},
+      target,
+      methodName,
+    );
+  };
+}
+
+export function getTestMetadata(
+  controllerClass: Constructor<{}>,
+  methodName: string,
+): {file: string} {
+  return Reflector.getMetadata(
+    'example.msg.decorator.metadata.key',
+    controllerClass.prototype,
+    methodName,
+  );
+}
+```
+
+**`test/unit/decorators/test.decorator.ts`**
+```ts
+import {test, getTestMetadata} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('test.decorator (unit)', () => {
+  it('can store test name via a decorator', () => {
+    class TestClass {
+      @test('me.test.ts')
+      me() {}
+    }
+
+    const metadata = getTestMetadata(TestClass, 'me');
+    expect(metadata).to.be.a.Object();
+    expect(metadata.file).to.be.eql('me.test.ts');
+  });
+});
+```
+
+#### Mixins
+
+A Mixin is a TypeScript function that extends the `Application` Class, adding
+new constructor properties, methods, etc. It is difficult to write a unit test
+for a Mixin without the `Application` Class dependency. The recommended practice
+is to write an integration test ofr a Mixin as show below.
+
+#### Providers
+
+A Provider is a Class that implements the Provider interface. This interface
+requires the Class to have a `value()` function. A unit test for a provider
+should test the `value()` function by instantiating a new Provider class, using
+a test double for any constructor arguments.
+
+**`src/providers/random-number.provider.ts`**
+```ts
+import {Provider} from '@loopback/context';
+
+export class RandomNumberProvider implements Provider<number> {
+  value(): number {
+    return (max: number): number => {
+      return Math.floor(Math.random() * max) + 1;
+    };
+  }
+}
+```
+
+**`test/unit/providers/random-number.unit.test.ts`**
+```ts
+import {RandomNumberProvider} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('RandomNumberProvider (unit)', () => {
+  it('generates a random number within range', () => {
+    const provider = new RandomNumberProvider().value();
+    const random: number = provider(3);
+
+    expect(random).to.be.a.Number();
+    expect(random).to.equalOneOf([1, 2, 3]);
+  })
+})
+```
+
+#### Repositories
+
+*TBD ... How / What?*
```

Comment:
This is a bit of a weird one. In our typical case, people will be using whatever repository implementation is provided by their utilities (either the way legacy-juggler builds Repository instances, or some other ORM).

In the case of [loopback4-extension-typeorm](https://github.com/strongloop/loopback4-extension-typeorm), it provides the functions you need to leverage the same concepts so that you don't have to build the repositories yourself as a user at runtime. What you'd test would be whether or not that binding is occurring, rather than how it gets made (in TypeORM's case, the _how_ is TypeORM's problem, not ours).

If an extension were to provide functions that directly produce objects that are "repositories", then that should be unit tested accordingly.

The TL;DR here is that if someone provides some form of repository functionality in their extension, they should either try to limit its responsibility to wiring up the repositories. If they can't, then they should write the usual unit and integration tests to prove that it works.

(Even talking about this stuff feels like it's something we shouldn't be writing an _opinion piece_ on, but I digress)

Reply:
From the readers' perspective, i think it's better to either remove the section or put "coming soon" (something like that) under it.   I'd actually prefer to remove it and create a separate ticket to track the outstanding items (whether we will write later or we decide we should write or not).